### ISSUE TITLE
F5 timeout issue with connectHandler

### ISSUE THREAD

**Issue by Trasmontinho** [ktbyers/netmiko#1599]
Hello, 

I've got in project to modify interfaces description depending on lldp from different vendors and kind of equipment. 

For this reason, I've seen a good opportunity to use your great tool. 

I've already prepared and tested with Cisco & Nokia successfully. I'm trying to use ConnectHandler to connect to F5 equipment in tmsh. I've been able to use the autodetect function that has determined & found F5 type. 

But when I've used connecthandler I faced a timeout issue (attached the output I had in my terminal)

I post also the logs activated into the script & my code. 

Thank you for your help, 

Regards,
[output_afterlaunching_script.txt](https://github.com/ktbyers/netmiko/files/4287705/output_afterlaunching_script.txt)
[logaction.txt](https://github.com/ktbyers/netmiko/files/4287714/logaction.txt)
[parsingscript.txt](https://github.com/ktbyers/netmiko/files/4287712/parsingscript.txt)

 
Sam

**Comment by Trasmontinho** [comment]
I've forgotten to give my version :-) It's the 3.0.0 and I'm using python 3.6.

Regards, 
Sam

**Comment by ktbyers** [comment]
Does F5 have a way to modify the terminal width? 

It looks like we are failing here:

```
  File "/usr/local/lib/python3.6/dist-packages/netmiko/f5/f5_tmsh_ssh.py", line 13, in session_preparation
    command="modify cli preference pager disabled display-threshold 0"
  File "/usr/local/lib/python3.6/dist-packages/netmiko/base_connection.py", line 1022, in disable_paging
    output = self.read_until_pattern(pattern=re.escape(command.strip()))
```

You can also probably fix this by downgrading to Netmiko 2.4.2 (but we should find a long-term fix in Netmiko 3.X so it would help if we can keep working on the underlying cause i.e. the question pertaining to terminal width.

**Comment by Trasmontinho** [comment]
Hello Kirk, 

I can help on the subject anytime. I still not understanding why because the two commands that are defined in "/usr/local/lib/python3.6/dist-packages/netmiko/f5/f5_tmsh_ssh.py" are accepted by the Big-ip. 

I'm not going to downgrade right now. I can skip for the moment the f5 stuff. Let me know if you need me to run some testing & debugging on it. 

Sam

**Comment by Trasmontinho** [comment]
Hello Kirk, 

I've troubleshooting and find maybe the issue. 

It seems that in base_connection, the function  _read_channel_expect(), we never match the condition at line 558 :
            if re.search(pattern, output, flags=re_flags):
                log.debug(f"Pattern found: {pattern} {output}")
                return output

I've added some log debug in this function (see attached in logdebug_added.txt) to show that. the debug output has shown that we bind the command and the prompt at the output string in two lines. the re.search never match the prompt in the output variable. 

Can you confirm that I'm right  ? 


Regards, 

Sam
[debugendoutput.txt](https://github.com/ktbyers/netmiko/files/4312056/debugendoutput.txt)
[logdebug_added.txt](https://github.com/ktbyers/netmiko/files/4312057/logdebug_added.txt)

**Comment by ktbyers** [comment]
Looking at your "debugendoutput.txt" file it looks like there are backspaces in the output:

```
DEBUG:netmiko:output: modify cli prefere ^Hnce pager disabled display-threshold 0
```

That extra space and ^H are a space and then a backspace (at least that is what control-H turned up for me).

So I would think it is probable this is the cause of the issue.


Not sure I understood your statement about the "bind the command and the prompt in the output string"...but I also doubt that is the issue.

**Comment by Trasmontinho** [comment]
Hello Kirk, 

I've looked to f5_tmsh_ssh.py and there's no extra character into the command. it seems that in that case the extra-character is introduced here "new_data = self.remote_conn.recv(MAX_BUFFER)" since before I haven't seen anything telling me that the command is transformed with this extra character.

When I'm telling "bind the command" it's because in the lasts debug output I've seen : 
DEBUG:netmiko:output: modify cli prefere nce pager disabled display-threshold 0
login@(F5)(cfg-sync In Sync)(Active)(/Common)(tmos)#  and type <class 'str'>

I've seen that in output the command is referenced + the prompt which is added after a second round in the while loop.

But as you said it doesn't seem to be the root cause.

Sam

**Comment by EnBee66** [comment]
Hello Kirk,

I have the same issue on F5 devices running on 12.1.5

Getting data from device types f5_ltm or f5_tmsh (there seems to be no difference in these two types) fails and the debug output shows 

DEBUG:netmiko:write_channel: b'modify cli preference pager disabled display-threshold 0\n'
DEBUG:netmiko:Pattern is: modify\ cli\ preference\ pager\ disabled\ display\-threshold\ 0
DEBUG:netmiko:_read_channel_expect read_data: modify cli pr eference pager disabled display-threshold 0

DEBUG:netmiko:write_channel: b'run /util bash -c "stty cols 255"\n'
DEBUG:netmiko:Pattern is: run\ /util\ bash\ \-c\ "stty\ cols\ 255"
DEBUG:netmiko:_read_channel_expect read_data: run /util bas h -c "stty cols 255"

The read data contains "space backspace", hex 20 and hex 08, at the same position in the output of both commands: 

modify cli pr eference pager disabled display-threshold 0
run /util bas h -c "stty cols 255"

Netmiko works when I modify f5_tmsh_ssh.py and set the actual commands for self.disable_paging and self.clear_buffer to "" (is no command at all)

Without "modify cli preference pager disabled display-threshold 0" commands with large output will fail. 
Adding this command as one of the commands in the script has the same result. If the script runs the commands "list ltm ntp" and then "modify cli preference pager disabled display-threshold 0", the first command will give the expected result. The second command will fail.

My workaround for this issue is adding a replace command in base_connection.py and leave file f5_tmsh_ssh.py as is.

# Changed this 
log.debug(f"_read_channel_expect read_data: {new_data}")
output += new_data


# To this
log.debug(f"_read_channel_expect read_data: {new_data}")
#  Used to filter "space - backspace char" in the result of some F5 commands
new_data = (new_data.replace('\x20\x08', ''))   
output += new_data



I know F5 support is experimental and may have some issues. One of the issues I have is that remote_conn.send_command(logcommand) fails most of the time for commands with a large amount of data.
It looks like Netmiko is not always able to "see" the expected data. The prompt of an F5 changes when the config of an HA pair in not in sync or when the active/standby mode changes. 
Netmiko has the option to set the expect_string for these kind of changes: 

remote_conn.send_command(logcommand, expect_string=prompt_to_expect   

For the F5 this is no guarantee Netmiko continues after the command is executed. 
To get the output of all commands I now use remote_conn.send_command_timing(logcommand, delay_factor=10))

It works, but without having a good explanation for this. 

I hope this info helps making Netmiko an even better tool.

**Comment by ktbyers** [comment]
Is someone available to work on this issue? I just don't have an F5 device available so I need someone that has a device available.

**Comment by EnBee66** [comment]
Hi Kirk, 

The company I work for has a test environment with F5 devices which I can use to perform some tests on. Please let me know how I can get useful information, then I will try to collect and share this with you.

**Comment by ktbyers** [comment]
@EnBee66 So I think we should start by trying to run the Netmiko tests (both show and config operations). I say this as I suspect there are some significant issues with the F5 driver, but it would be good to get an overall picture of what is working and not working.

The test process is detailed here:

https://github.com/ktbyers/netmiko/blob/develop/TESTING.md

Let me know if this is something you are willing to do and if you need any assistance on this.

If you are able to do this, then just post the test output here (at least as far as passed/failed tests and we can then proceed from there).

**Comment by EnBee66** [comment]
Hi Kirk,
After reading the instructions I have tried to understand what I have to add in the files 
commands.yml and responses.yml. Until now I was not able to get output from the script test_netmiko_show.py, mainly because I can't find the file py.test. 

Netmiko is able to detect the correct device type for the the accounts I can use to access the CLI. 
The paging commands also work when using the work around I discribed. 

## username is a personal account (only TMSH access)
## c:\data1\nm-aut>autodetect.py
## password:
## best_match: f5_tmsh
## {'f5_tmsh': 99}
## <class 'str'>

## username is root, CLI access (aka advanced shell on F5 devices)
## c:\data1\nm-aut>autodetect.py
## password:
## best_match: f5_linux
## {'f5_linux': 99}
## <class 'str'> 

If you can give me some hints on how to perform the tests and which result you expect, I will try to get it for you. Thanks for your time and effort. 

Nico

**Comment by EnBee66** [comment]
FYI:

I was not able to execute the test process but I did notice some changes in the base_connection.py file of the Netmiko development branch. If you don't try you will not learn, so I copied this file over the current base_connection.py file and this solves the issue we have with having the "space backspace" charachters in the output of  netmiko:_read_channel_expect read_data. 


**!# base_connection.py  - file from current version 3.1.0**
DEBUG:netmiko:write_channel: b'modify cli preference pager disabled display-threshold 0\n'
DEBUG:netmiko:Pattern is: modify\ cli\ preference\ pager\ disabled\ display\-threshold\ 0
DEBUG:netmiko:_read_channel_expect read_data: modify _cli pr eference_ pager disabled display-threshold 0

**!# base_connection.py  - file from developement version**
DEBUG:netmiko:write_channel: b'modify cli preference pager disabled display-threshold 0\n'
DEBUG:netmiko:Pattern is: modify\ cli\ preference\ pager\ disabled\ display\-threshold\ 0
DEBUG:netmiko:_read_channel_expect read_data: modify _cli preference_ pager disabled display-threshold 0

**Comment by ktbyers** [comment]
@EnBee66 Which line did you modify?

**Comment by EnBee66** [comment]
Hi Kirk,

In the original file base_connection.py of version 3.1.0 I added one line. This line removes the "space backspace" combination. 

!# Changed this
log.debug(f"_read_channel_expect read_data: {new_data}")
output += new_data

!# To this
log.debug(f"_read_channel_expect read_data: {new_data}")
!# Used to filter "space - backspace char" in the result of some F5 commands
**_new_data = (new_data.replace('\x20\x08', ''))_**
output += new_data

In my last post I try to explain that I use all files from version 3.1.0 and only replaced the file base_connection.py with the file base_connection.py from the development branch without further editing of that file.   The space backspace combination is seen in the output that is returned by Netmiko (bold and Italic in the output below) This is the command that was executed.


**!# base_connection.py - file from current version 3.1.0**
!##############################################################
- Command: show running-config all-properties recursive
**_show running-config al l-properties recursive_**
analytics gui-widget /Common/uwid5a0abacd8c191 {


The file from the dev version does not return the command executed. 
!# base_connection.py - file from developement version
!##############################################################
- Command: show running-config all-properties recursive
<< this line is blank >>
analytics gui-widget /Common/uwid5a0abacd8c191 {

This still does not explain what causes of the "space backspace" characters Netmiko shows in the output and the issue can return in future versions of Netmiko. 
I have added a zip file with three versions of the file base_connection.py. The original from v3.1.0, the one from v3.1.0 with the filter for space backspace (works also with Cisco devices btw) and the file from the current development version. 


[base_connection_versions.zip](https://github.com/ktbyers/netmiko/files/4588384/base_connection_versions.zip)

**Comment by ktbyers** [comment]
It is the F5 that is adding the space-backspace (not Netmiko). Netmiko is just trying to workaround what the F5 does.

Usually this happens as the network device is trying to be smart about formatting the output to make it more readable for people (but of course we have no people directly looking at the terminal-output here).

So it works using the `development` base_connection.py?

**Comment by EnBee66** [comment]
Q - So it works using the develop base_connection.py?
A - Yes

I asume this is because the command that was send to the F5 device is not expected back in the read channel. The result of the command in the write channel is different from the command in the read channel.  It then contains the "space backspace" characters.  Maybe the F5 indeed inserts these characters.   

This is the debug result of the original base_connection.py

DEBUG:netmiko:Command: modify cli preference pager disabled display-threshold 0
DEBUG:netmiko:write_channel: b'modify cli preference pager disabled display-threshold 0\n'
DEBUG:netmiko:Pattern is: modify\ cli\ preference\ pager\ disabled\ display\-threshold\ 0
DEBUG:netmiko:_read_channel_expect read_data: modify  cli preference pager disabled display-threshold 0 

As mentioned, the base_connection.py seems not to look at the data from the read channel. 


Is there a way to see who inserts the space backspace characters? (Eg, can I dump the output of the read channel)

**Comment by ktbyers** [comment]
You can see in the above output that it is the F5 that is modifying it:

Here is what Netmiko sent:

```
DEBUG:netmiko:write_channel: b'modify cli preference pager disabled display-threshold 0\n'
```

So Netmiko is writing out:

```
modify cli preference pager disabled display-threshold 0
```

And here is the echoed command coming back from the F5:

```
modify � cli preference pager disabled display-threshold 0
```

That space-question-mark is from the F5. That `read_channel` logging is very low-level (i.e. it is pretty much right when we read it back from the device).

I think it is probably this PR here that fixed the issue.

https://github.com/ktbyers/netmiko/pull/1673

In other words, I recently put in a fix that sets the default terminal width to be wider (and F5 is doing the above modification because it thinks the terminal is narrow (and thus that it needs to adjust it to make it more human-readable).

Anyways if it works in the `develop` branch, then you should just use that until the next official Netmiko release.

**Comment by EnBee66** [comment]
I can confirm the change from:

    `def establish_connection(self, width=None, height=None):`
to
    `def establish_connection(self, width=511, height=1000):`

indeed makes the difference. 

I have tested a connection directly using paramiko and setting the width to at least 127 makes the space backspace sequence disappear.  (  ssh_client.invoke_shell(width=127)  )

One last question: Can I set the terminal width that Netmiko will use as a parameter in my script?
If so, my colleagues  can use the standard netmiko package. If this is not possible we have to use the file from the development version. 

Thanks for your replies
